<?xml version="1.0" encoding="UTF-8"?>

<mule xmlns:dw="http://www.mulesoft.org/schema/mule/ee/dw" xmlns:tracking="http://www.mulesoft.org/schema/mule/ee/tracking" xmlns="http://www.mulesoft.org/schema/mule/core" xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
	xmlns:spring="http://www.springframework.org/schema/beans" 
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-current.xsd
http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/ee/tracking http://www.mulesoft.org/schema/mule/ee/tracking/current/mule-tracking-ee.xsd
http://www.mulesoft.org/schema/mule/ee/dw http://www.mulesoft.org/schema/mule/ee/dw/current/dw.xsd">
    
    <choice-exception-strategy name="global-exception-strategy">    
        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.http.internal.request.ResponseValidatorException)]" doc:name="Caused By (org.mule.module.http.internal.request.ResponseValidatorException)">
            <choice doc:name="Choice">
                <when expression="#[message.inboundProperties['http.status'] == 400]">
                    <dw:transform-message doc:name="Transform Message">
                        <dw:set-variable variableName="errorMessage"><![CDATA[%dw 1.0
%output application/java
---
payload
]]></dw:set-variable>
                    </dw:transform-message>
                    <flow-ref name="global-bad-request-response-sub-flow" doc:name="Bad Gateway Response"/>

                </when>
                <when expression="#[message.inboundProperties['http.status'] == 401]">
                    <set-variable variableName="errorMessage" value="Upstream service did not authorize the request." doc:name="Set Error Message"/>
                    <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="global-bad-gateway-response-sub-flow"/>
                </when>
                <when expression="#[message.inboundProperties['http.status'] == 403]">
                    <set-variable variableName="errorMessage" value="Access to the upstream service is forbidden." doc:name="Set Error Message"/>
                    <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="Bad Gateway Response"/>
                </when>
                <when expression="#[message.inboundProperties['http.status'] &lt; 500]">
                    <logger message="This is what we got back from a 500 error: #[message.payloadAs(java.lang.String)]" level="ERROR" category="com.merrillcorp.logging" doc:name="Logger"/>
                    <set-variable variableName="errorMessage" value="Upstream service unable to fulfil request." doc:name="Set Error Message"/>
                    <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="Bad Gateway Response"/>
                </when>
                <otherwise>
                    <set-variable variableName="errorMessage" value="Upstream service internal error." doc:name="Set Error Message"/>
                    <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="global-bad-gateway-response-sub-flow"/>
                </otherwise>
            </choice>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(java.util.concurrent.TimeoutException)]" doc:name="Caused By (java.util.concurrent.TimeoutException)">
            <set-variable variableName="errorMessage" value="Unable to connect to upstream service. Request timed out." doc:name="Set Error Message"/>
            <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="Bad Gateway Response"/>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(java.util.NoSuchElementException)]" doc:name="Caused By (java.util.NoSuchElementException)">
            <set-variable variableName="errorMessage" value="Unable to connect to upstream service." doc:name="Set Error Message"/>
            <flow-ref name="global-bad-gateway-response-sub-flow" doc:name="Bad Gateway Response"/>
        </catch-exception-strategy>

        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.apikit.exception.NotFoundException)]" doc:name="Caused By (org.mule.module.apikit.exception.NotFoundException)">
            <flow-ref name="global-resource-not-found-response-sub-flow" doc:name="Resource Not Found Response"/>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.apikit.exception.MethodNotAllowedException)]" doc:name="Caused By (org.mule.module.apikit.exception.MethodNotAllowedException)">
            <flow-ref name="global-method-not-allowed-response-sub-flow" doc:name="Method Not Allowed Response"/>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.apikit.exception.UnsupportedMediaTypeException)]" doc:name="Caused By (org.mule.module.apikit.exception.UnsupportedMediaTypeException)">
            <flow-ref name="global-unsupported-media-type-response-sub-flow" doc:name="Unsupported Media Type Response"/>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.apikit.exception.NotAcceptableException)]" doc:name="Caused By (org.mule.module.apikit.exception.NotAcceptableException)">
            <flow-ref name="global-not-acceptable-response-sub-flow" doc:name="Not Acceptable Response"/>
        </catch-exception-strategy>
        <catch-exception-strategy when="#[exception.causedBy(org.mule.module.apikit.exception.BadRequestException)]" doc:name="Caused By (org.mule.module.apikit.exception.BadRequestException)">
            <flow-ref name="global-bad-request-response-sub-flow" doc:name="Bad Request Response"/>
        </catch-exception-strategy>
        
        <catch-exception-strategy when="#[exception!=null &amp;&amp; (exception.causedBy(java.io.IOException) || exception.causedBy(java.net.ConnectException)  || exception.causedBy(java.net.SocketTimeoutException))]" doc:name="Caused after untill successful gets exhausted.">
            <flow-ref name="global-until-successful-sub-flow" doc:name="global-until-successful-sub-flow"/>
        </catch-exception-strategy>
        
        <catch-exception-strategy doc:name="Default">
            <set-variable variableName="errorMessage" value="Unable to fullfil request due to internal error." doc:name="Set Error Message"/>
            <flow-ref name="global-internal-server-error-response-sub-flow" doc:name="global-internal-server-error-response-sub-flow"/>
        </catch-exception-strategy>
    </choice-exception-strategy>
    
    <sub-flow name="global-bad-request-response-sub-flow">
        <set-property propertyName="http.status" value="400" doc:name="Set Status"/>
        <set-payload value="BAD_REQUEST" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="Prepare Error Response"/>
    </sub-flow>
    <sub-flow name="global-resource-not-found-response-sub-flow">
        <set-property propertyName="http.status" value="404" doc:name="Set Status"/>
        <set-payload value="RESOURCE_NOT_FOUND" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="global-prepare-error-response-sub-flow"/>
    </sub-flow>
    <sub-flow name="global-method-not-allowed-response-sub-flow">
        <set-property propertyName="http.status" value="405" doc:name="Set Status"/>
        <set-payload value="METHOD_NOT_ALLOWED" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="Prepare Error Response"/>
    </sub-flow>
    <sub-flow name="global-not-acceptable-response-sub-flow">
        <set-property propertyName="http.status" value="406" doc:name="Set Status"/>
        <set-payload value="NOT_ACCEPTABLE" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="Prepare Error Response"/>
    </sub-flow>
    <sub-flow name="global-conflict-response-sub-flow">
        <set-property propertyName="http.status" value="409" doc:name="Set Status"/>
        <set-payload value="CONFLICT" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="Prepare Error Response"/>
    </sub-flow>
    <sub-flow name="global-unsupported-media-type-response-sub-flow">
        <set-property propertyName="http.status" value="415" doc:name="Set Status"/>
        <set-payload value="UNSUPPORTED_MEDIA_TYPE" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="Prepare Error Response"/>
    </sub-flow>
    <sub-flow name="global-internal-server-error-response-sub-flow">
        <set-property propertyName="http.status" value="500" doc:name="Set Status"/>
        <set-payload value="INTERNAL_SERVER_ERROR" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="global-prepare-error-response-sub-flow"/>
    </sub-flow>
    <sub-flow name="global-bad-gateway-response-sub-flow">
        <set-property propertyName="http.status" value="502" doc:name="Set Status"/>
        <set-payload value="BAD_GATEWAY" doc:name="Set Error Code" mimeType="application/java"/>
        <flow-ref name="global-prepare-error-response-sub-flow" doc:name="global-prepare-error-response-sub-flow"/>
    </sub-flow>
    
    <sub-flow name="global-prepare-error-response-sub-flow">
        <logger message="Transaction [#[flowVars.transactionId]] - Error Code [#[payload]] - Error Message [#[flowVars.errorMessage != null ? flowVars.errorMessage : '']] - Error Description [#[flowVars.errorDescription != null ? flowVars.errorDescription : '']]" level="ERROR" category="com.merrillcorp.q2c" doc:name="Logger"/>
        <set-variable variableName="errorRaised" value="#[true]" doc:name="Set Error Raised"/>
    </sub-flow>
    
    <sub-flow name="global-until-successful-sub-flow">
        <logger message="#[sessionVars.payloadBeforeUntilSuccess]" level="ERROR" category="com.merrillcorp.q2c.retry" doc:name="Logger"/>
    </sub-flow>
    
</mule>